oeps/oep-0006.rst

```diff
@@ -0,0 +1,410 @@
+
+  * This scope has been dropped from the proposal as it seems to lack a clear
+    and compelling use case.
+
+* ``Scope.context_shared_settings`` (UserScope.NONE, BlockScope.CONTEXT_ALL_TYPES)
+
+  * This scope has been dropped from the proposal as it seems to lack a clear
+    and compelling use case; further, it is unclear how an authoring experience
+    for such fields would work, since no one XBlock could provide an authoring
+    UI for fields that could be shared by many unrelated XBlocks.
+
+
+Inheritance
+-----------
+
+The current proposal deliberately does not support inheritance (i.e. a "context"
+in the LMS would always be a course, not a "section" or any smaller unit, and
+field values don't propagate from parent blocks to children), so it is not a
+replacement for `InheritanceMixin`_.
+
+Initial discussions of this OEP agreed that context-scoped fields should not
+participate in inheritance. Improving inheritance of fields or making
+inheritable fields pluggable are worthy goals, but for a separate proposal.
+
+.. _InheritanceMixin: https://github.com/edx/edx-platform/blob/46d69eba7ac45/common/lib/xmodule/xmodule/modulestore/inheritance.py#L30
+
+Context as a New Scope Dimension
+--------------------------------
+
+Currently, a ``BlockScope``, a ``UserScope``, and a ``name`` are combined to
+make a Scope for a field. Under this alternative option, a new
+``RelevanceScope`` dimension would be added, which could have one of two values:
+``ALL`` (default) or ``CONTEXT``. The ``ALL`` value would represent the current
+XBlock API's definitions, and the ``CONTEXT`` value would mean that the field's
+relevance is limited to one particular context (e.g. that field's value relates
+to one particular course.)
+
+The downside of this is that it creates a large matrix/tensor of scopes:
+4 BlockScopes × 3 UserScopes × 2 RelevanceScopes = 24 distinct scope types.
+Not only is this more difficult to understand, but there is no value in defining
+a "context-specific" version of any ``BlockScope.USAGE`` or
+``BlockScope.DEFINITION`` fields. As a result of that and other conflicting or
+not useful Scope combinations, it is likely that fewer than half of the 24
+possible Scope values would be used.
+
+Shared data service
+-------------------
+
+Another alternate option is to create a "shared data" or "course-scoped data"
+XBlock service that XBlocks can use to read and write data at a course level,
+rather than making this a new type of field. Such an approach seems less
+consistent, and would not allow the automatic creationn of a UI for editing
+context-scoped XBlock settings. (With the field approach, we can replace the
+"Advanced Settings" page in Studio with an auto-generated UI that allows authors
+to configure the course-scoped settings of all XBlocks enabled for that course.
+We can also generate a "preferences" tab in the LMS that allows students to
+change XBlock preferences for that course.)
```

Comment:
This is good food for thought as we think through all the scenarios required for Discussions, including its possible future evolution.